frontend/gateway/grpcclient/client.go

```diff
@@ -47,10 +48,98 @@ func Current() (client.Client, error) {
 		opts:      opts(),
 		sessionID: sessionID(),
 		workers:   workers(),
+		product:   product(),
 		caps:      pb.Caps.CapSet(resp.FrontendAPICaps),
+		requests:  map[string]*pb.SolveRequest{},
 	}, nil
 }
 
+func convertRefs(refs map[string]client.Reference) (map[string]string, error) {
+	out := make(map[string]string, len(refs))
+	for k, v := range refs {
+		if v == nil {
+			out[k] = ""
+		} else {
+			r, ok := v.(*reference)
+			if !ok {
+				return nil, errors.Errorf("invalid return reference type %T", v)
+			}
+			out[k] = r.id
+		}
+	}
+	return out, nil
+}
+
+func Run(ctx context.Context, f client.BuildFunc) (retError error) {
+	client, err := current()
+	if err != nil {
+		return errors.Wrapf(err, "failed to initialize client from environment")
+	}
+
+	res, err := f(ctx, client)
+
+	export := client.caps.Supports(pb.CapReturnResult) == nil
+
+	if export {
+		defer func() {
+			req := &pb.ReturnRequest{}
+			if retError == nil {
+				refs, err := convertRefs(res.Refs)
```

Comment:
This would panic if `res` is nil, which could be the case since `err` from the build func is checked later.

Reply:
this is in a defer

Reply:
Actually the fact that this is under a `retError == nil` case makes it less likely to cause problems.
If build func returns an error, `res.Refs` is skipped, and if it returned no errors, then **by convention**, `res` should not be nil.